test/unit/assorted/server_discovery_and_monitoring.spec.test.ts

```diff
@@ -0,0 +1,513 @@
+        } else {
+          // unset or true means no errors should be thrown
+          expect(errorsThrown).to.be.empty;
+        }
+      } else {
+        expect.fail(ejson`Unknown outcome shape - ${phase.outcome}`);
+      }
+
+      events = [];
+      errorsThrown = [];
+    }
+  }
+}
+
+function withConnectionStubImpl(appError) {
+  return function (conn, fn, callback) {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const connectionPool = this; // we are stubbing `withConnection` on the `ConnectionPool` class
+    const fakeConnection = {
+      generation:
+        typeof appError.generation === 'number' ? appError.generation : connectionPool.generation,
+
+      command(ns, cmd, options, callback) {
+        if (appError.type === 'network') {
+          callback(new MongoNetworkError('test generated'));
+        } else if (appError.type === 'timeout') {
+          callback(
+            new MongoNetworkTimeoutError('xxx timed out', {
+              beforeHandshake: appError.when === 'beforeHandshakeCompletes'
+            })
+          );
+        } else {
+          callback(new MongoServerError(appError.response));
+        }
+      }
+    };
+
+    fn(undefined, fakeConnection, (fnErr, result) => {
+      if (typeof callback === 'function') {
+        if (fnErr) {
+          callback(fnErr);
+        } else {
+          callback(undefined, result);
+        }
+      }
+    });
+  };
+}
+
+function assertTopologyDescriptionOutcomeExpectations(
+  topology: Topology,
+  outcome: TopologyDescriptionOutcome
+) {
+  assertTopologyDescriptionOutcome(outcome);
+  // then verify the resulting outcome
+  const description = topology.description;
+
+  expect(description).to.have.property('type', outcome.topologyType);
+
+  const expectedServers = new Map(Object.entries(outcome.servers));
+  const actualServers = description.servers;
+  expect(actualServers).to.be.instanceOf(Map);
+
+  // TODO(NODE-4159): The node driver keeps unknown servers where it should discard
+  // expect(actualServers.size).to.equal(expectedServers.size);
+
+  for (const serverName of expectedServers.keys()) {
+    expect(actualServers).to.include.keys(serverName);
+    const expectedServer = expectedServers.get(serverName);
+
+    if (expectedServer.pool != null) {
+      const expectedPool = expectedServers.get(serverName).pool;
+      const actualPoolGeneration = topology.s.servers.get(serverName).s.pool;
+      expect(actualPoolGeneration).to.have.property('generation', expectedPool.generation);
+      delete expectedServer.pool;
+    }
+
+    const normalizedExpectedServer = normalizeServerDescription(expectedServer);
+    const omittedFields = findOmittedFields(normalizedExpectedServer);
+    const actualServer = actualServers.get(serverName);
+    expect(actualServer).to.matchMongoSpec(normalizedExpectedServer);
+
+    if (omittedFields.length !== 0) {
+      expect(actualServer).to.not.have.all.keys(omittedFields);
```

Comment:
this should be `any` not `all`, because using an "all" in the negation would pass if some omittedFields were still present. I strongly recommend referring to the `keys` [chai docs](https://www.chaijs.com/api/bdd/#method_keys) whenever you use `any` or `all` because it's very easy to get wrong

Reply:
Yea this assertion was here before but I missed it, looks like we have some bugs if this assertion is necessary. 
For example "Non-stale network error" in `test/spec/server-discovery-and-monitoring/errors/non-stale-network-error.json` expects topologyVersion to be null, because the server is marked unknown, I'm adding this to NODE-4159 I think it directly relates.

Reply:
yay for catching errors by using correct assertions, lol
maybe instead of leaving this assertion in the weird state, we could do a `for .. of` on the `omittedFields` and expect `.to.not.have.property` with exceptions for the known incompatibilities?

Reply:
Leaving this open for context for other reviewers, can be resolved if everyone approves